lib/class_def_ast.js

```diff
@@ -322,6 +413,163 @@ class ClassDef {
+                auth['discoveryType'] = param.value.value;
+            else if (param.value.isString)
+                auth[param.name] = getString(param.value);
+            else
+                auth[param.name] = param.value;
+        });
+        if (this.config) {
+            switch (this.config.module) {
+                case 'org.thingpedia.config.oauth2':
+                    auth.type = 'oauth2';
+                    break;
+                case 'org.thingpedia.config.custom_oauth':
+                    auth.type = 'custom_oauth';
+                    break;
+                case 'org.thingpedia.config.basic_auth':
+                    auth.type = 'basic';
+                    break;
+                case 'org.thingpedia.config.discovery':
+                    auth.type = 'discovery';
+                    break;
+                case 'org.thingpedia.config.interactive':
+                    auth.type = 'interactive';
+                    break;
+                case 'org.thingpedia.config.builtin':
+                    auth.type = 'builtin';
+                    break;
+                default:
+                    auth.type = 'none';
+            }
+        }
+        return auth;
+    }
+
+    toManifest() {
+        let [queries, actions] = [{}, {}];
+        Object.entries(this.queries).forEach(([name, query]) => queries[name] = query.toManifest());
+        Object.entries(this.actions).forEach(([name, action]) => actions[name] = action.toManifest());
+        let manifest = {
+            module_type: this.loader ? this.loader.module : 'org.thingpedia.v2',
+            kind: this.kind,
+            category: getString(this.annotations.category),
+            subcategory: getString(this.annotations.subcategory),
+            params: this._params(),
+            auth: this._auth(),
+            queries,
+            actions,
+            version: this.annotations.version ? this.annotations.version.value : undefined,
+            types: this.extends,
+            child_types: this.annotations.child_types ? this.annotations.child_types.value.map((v) => getString(v)) : []
+        };
+        if (this.metadata.name) manifest.name = this.metadata.name;
+        if (this.metadata.description) manifest.description = this.metadata.description;
+        return manifest;
+    }
+
+    static fromManifest(kind, manifest) {
+        let _extends = manifest.types || [];
+        let imports = [];
+        let queries = {};
+        let actions = {};
+        let metadata = {};
+        if (manifest.name)
+            metadata.name = Ast.Value.String(manifest.name);
+        if (manifest.description)
+            metadata.description = Ast.Value.String(manifest.description);
+        let annotations = {
+            category: Ast.Value.String(manifest.category),
+            subcategory: Ast.Value.String(manifest.subcategory),
+            version: manifest.version ? Ast.Value.Number(manifest.version) : undefined,
```

Comment:
Actually this is buggy: version === 0 is a valid version, but 0 is false in JavaScript.